geo/src/algorithm/line_interpolate_point.rs

```diff
@@ -1,7 +1,137 @@
+use geo_types::{MultiLineString, MultiPolygon};
+
 use crate::coords_iter::CoordsIter;
 use std::ops::AddAssign;
 
-use crate::{CoordFloat, EuclideanLength, Line, LineString, Point};
+use crate::{CoordFloat, EuclideanLength, Line, LineString, Point, Polygon};
+
+/// Return a new linear geometry containing both existing and new interpolated coordinates with
+/// a maximum distance of `max_distance` between them.
+///
+/// # Examples
+/// ```
+/// use geo::{coord, Line, LineString};
+/// use geo::line_interpolate_point::Densify;
+///
+/// let line: Line<f64> = Line::new(coord! {x: 0.0, y: 6.0}, coord! {x: 1.0, y: 8.0});
+/// let correct: LineString<f64> = vec![[0.0, 6.0], [0.5, 7.0], [1.0, 8.0]].into();
+/// let max_dist = 2.0;
+/// let densified = line.densify(max_dist);
+/// assert_eq!(densified, correct);
+///```
+pub trait Densify<F: CoordFloat> {
+    type Output;
+
+    fn densify(&self, max_distance: F) -> Self::Output;
+}
+
+// Helper for densification trait
+fn densify_line<T: CoordFloat>(line: Line<T>, container: &mut Vec<Point<T>>, max_distance: T) {
+    container.push(line.start_point());
+    let num_segments = (line.euclidean_length() / max_distance).ceil() - T::one();
```

Comment:
Some more take it or leave it feedback:

I think like the `+ T::one()`'s below can go away if we remove this `- T::one()` and adjust the iteration... something like:

```
    let num_segments = (line.euclidean_length() / max_distance).ceil().to_u64().unwrap();
    // distance "unit" for this line segment
    let frac = T::one() / T::from(num_segments).unwrap();

    for segment_idx in 1..num_segments {
        let ratio = frac * T::from(segment_idx).unwrap();
        let interpolated_point = line.line_interpolate_point(ratio).expect("ratio should be between 0..1");
        container.push(interpolated_point);
    }
```

I think what you have here is correct, it just seems like it would be simpler to understand without the adjustments.

Reply:
🤦‍♂️ I made this way too complicated while trying to get the logic correct